homeassistant/components/recorder/__init__.py

```diff
@@ -314,127 +352,198 @@ def do_adhoc_purge(self, **kwargs):
+            try:
+                self._process_one_event_or_recover(event)
+            except Exception as err:  # pylint: disable=broad-except
+                _LOGGER.exception("Error while processing event %s: %s", event, err)
 
-            if event is None:
-                self._shutdown()
-                return
+        self._shutdown()
 
+    def _process_one_event_or_recover(self, event):
+        """Process an event, reconnect, or recover a malformed database."""
+        try:
             self._process_one_event(event)
+            return
+        except exc.DatabaseError as err:
+            if not self._handle_database_error(err):
+                _LOGGER.exception(
+                    "Database error while processing event %s: %s", event, err
+                )
+        except SQLAlchemyError as err:
+            _LOGGER.exception(
+                "SQLAlchemyError error processing event %s: %s", event, err
+            )
 
-    def _setup_recorder(self) -> bool:
-        """Create schema and connect to the database."""
+        # Reset the session if an SQLAlchemyError (including DatabaseError)
+        # happens to rollback and recover
+        self._reopen_event_session()
+
+    def _setup_recorder(self) -> None | int:
+        """Create connect to the database and get the schema version."""
         tries = 1
 
         while tries <= self.db_max_retries:
             try:
                 self._setup_connection()
-                migration.migrate_schema(self)
-                self._setup_run()
+                return migration.get_schema_version(self)
             except Exception as err:  # pylint: disable=broad-except
                 _LOGGER.exception(
                     "Error during connection setup to %s: %s (retrying in %s seconds)",
                     self.db_url,
                     err,
                     self.db_retry_wait,
                 )
-            else:
-                _LOGGER.debug("Connected to recorder database")
-                self._open_event_session()
-                return True
-
             tries += 1
             time.sleep(self.db_retry_wait)
 
-        @callback
-        def connection_failed():
-            """Connect failed tasks."""
-            self.async_db_ready.set_result(False)
-            persistent_notification.async_create(
-                self.hass,
-                "The recorder could not start, please check the log",
-                "Recorder",
-            )
+        return None
 
-        self.hass.add_job(connection_failed)
-        return False
+    def _migrate_schema_and_setup_run(self, current_version) -> bool:
+        """Migrate schema to the latest version."""
+        persistent_notification.create(
+            self.hass,
+            "System performance will temporarily degrade during the database upgrade. Integrations that read the database, such as logbook and history, may return inconsistent results until the upgrade completes.",
```

Comment:
Should we mention that they should not restart their Home Assistant installation?

We probably want to show this in the frontend somewhere too.

Reply:
Yes. That's a good idea 👍

Reply:
Maybe fire an event 'recorder_migration_started' 'recorder_migration_ended'. We could then show a warning on the server tools by the restart button that they probably shouldn't do it

Reply:
I wonder if shutdown will block actually since the thread won't exit until its finished.

Reply:
`System performance will temporarily degrade during the database upgrade. Do not power down or restart the system until the upgrade completes. Integrations that read the database, such as logbook and history, may return inconsistent results until the upgrade completes.`

Changed wording to ^

Reply:
The webserver will shutdown but the database migration continues and restart is blocked until it finishes because the thread won't join.

Reply:
2021-04-12 06:08:27 WARNING (MainThread) [homeassistant.core] Timed out waiting for shutdown stage 1 to complete, the shutdown will continue

Reply:
Shutdown blocks until the migration is complete because the thread can't join.  I'm going to do another PR to check if migration is in progress and throw when calling `SERVICE_HOMEASSISTANT_STOP` or `SERVICE_HOMEASSISTANT_RESTART` so they don't end up in this state unexpectedly.

```
Thread 213 (idle): "MainThread"
    _wait_for_tstate_lock (threading.py:1027)
    join (threading.py:1011)
    shutdown (concurrent/futures/thread.py:236)
    close (homeassistant/runner.py:92)
    run (asyncio/runners.py:51)
    run (homeassistant/runner.py:126)
    main (homeassistant/__main__.py:313)
    <module> (homeassistant/__main__.py:321)
    _run_code (runpy.py:87)
    _run_module_as_main (runpy.py:194)
Thread 237 (idle): "SyncWorker_1"
    _wait_for_tstate_lock (threading.py:1027)
    join (threading.py:1011)
    shutdown (homeassistant/components/recorder/__init__.py:364)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:932)
    _bootstrap (threading.py:890)
Thread 245 (idle): "Recorder"
    migrate_schema (homeassistant/components/recorder/migration.py:60)
    _migrate_schema_and_setup_run (homeassistant/components/recorder/__init__.py:523)
    run (homeassistant/components/recorder/__init__.py:435)
    _bootstrap_inner (threading.py:932)
    _bootstrap (threading.py:890)
Thread 251 (idle): "Thread-8"
    loop (paho/mqtt/client.py:1167)
    loop_forever (paho/mqtt/client.py:1779)
    _thread_main (paho/mqtt/client.py:3452)
    run (threading.py:870)
    _bootstrap_inner (threading.py:932)
    _bootstrap (threading.py:890)
Thread 259 (idle): "Thread-9"
    loop (paho/mqtt/client.py:1167)
    loop_forever (paho/mqtt/client.py:1779)
    _thread_main (paho/mqtt/client.py:3452)
    run (threading.py:870)
    _bootstrap_inner (threading.py:932)
    _bootstrap (threading.py:890)

````

Reply:
Verified everything went down cleanly and restarted successfully once the migration did finish.

Reply:
Service block started here https://github.com/home-assistant/core/pull/49098

Need to think about how to tell the frontend as it needs to know on reload.